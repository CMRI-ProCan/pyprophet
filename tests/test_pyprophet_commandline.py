import os
import subprocess
import shutil
import sys

import pandas as pd

import pytest


d = pd.options.display
d.width = 220
d.precision = 6

DATA_FOLDER = os.path.join(os.path.dirname(os.path.abspath(__file__)), "data")


def _record(stdout, regtest):
    for line in stdout.split("\n"):
        if line.startswith("INFO"):
            # remove data which is not constant over runs as time or process id:
            line = line.split(":", 4)[-1]
        if "TIME" in line:
            continue
        print >> regtest, line


def _expected_output_files():
    return ["test_data_ms2_scored.tsv",
            "test_data_ms2_report.pdf",
            "test_data_ms2_weights.csv",
            "test_data_ms2_summary_stat.csv",
            "test_data_ms2_full_stat.csv",
            ]


def _remove_output_files(tmpdir, names=None):
    if names is None:
        names = _expected_output_files()
    for name in names:
        full_path = os.path.join(tmpdir.strpath, name)
        os.remove(full_path)


def _dump_output_files(tmpdir, regtest, names=None):
    try:
        tmpdir = tmpdir.strpath
    except AttributeError:
        pass
    if names is None:
        names = _expected_output_files()
    for name in names:
        full_path = os.path.join(tmpdir, name)
        if not name.endswith(".pdf"):
            _dump(full_path, regtest)


def _dump(full_path, regtest):
    head = open(full_path).readline()
    if "\t" in head:
        df = pd.read_csv(full_path, sep="\t", header=None)
    else:
        df = pd.read_csv(full_path, sep=" ", header=None)

    # lines = open(full_path, "r").readlines()
    # f = os.path.basename(full_path)
    if len(df) > 20:
        print >> regtest, df[:10]
        print >> regtest, "..."
        print >> regtest, df[-10:]
    else:
        print >> regtest, df


def _run_cmdline(cmdline):
    stdout = cmdline + "\n"
    try:
        stdout += subprocess.check_output(cmdline, shell=True,
                                          stderr=subprocess.STDOUT)
    except subprocess.CalledProcessError, e:
        print >> sys.stderr, e.output
        raise
    return stdout


def _run_pyprophet_to_learn_model(regtest, temp_folder, dump_result_files=False, parametric=False, pfdr=False, pi0_lambda=False):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.txt")
    shutil.copy(data_path, temp_folder)
    cmdline = "pyprophet score --in=test_data.txt --test"
    if parametric:
        cmdline += " --parametric"
    if pfdr:
        cmdline += " --pfdr"
    if pi0_lambda is not False:
        cmdline += " --pi0_lambda=" + pi0_lambda
    stdout = _run_cmdline(cmdline)

    for f in _expected_output_files():
        full_path = os.path.join(temp_folder, f)
        assert os.path.exists(full_path)

    if not dump_result_files:
        return stdout

    _dump_output_files(temp_folder, regtest)

    return stdout


def test_tsv_0(tmpdir, regtest):
    stdout = _run_pyprophet_to_learn_model(regtest, tmpdir.strpath, True)
    _record(stdout, regtest)

def test_tsv_1(tmpdir, regtest):
    stdout = _run_pyprophet_to_learn_model(regtest, tmpdir.strpath, True, True)
    _record(stdout, regtest)

def test_tsv_2(tmpdir, regtest):
    stdout = _run_pyprophet_to_learn_model(regtest, tmpdir.strpath, True, False, True)
    _record(stdout, regtest)

def test_tsv_3(tmpdir, regtest):
    stdout = _run_pyprophet_to_learn_model(regtest, tmpdir.strpath, True, False, False, "0.3 0.55 0.05")
    _record(stdout, regtest)

def test_tsv_apply_weights(tmpdir, regtest):

    stdout = _run_pyprophet_to_learn_model(regtest, tmpdir.strpath, True)

    _record(stdout, regtest)

    full_path = os.path.join(tmpdir.strpath, "test_data_ms2_scored.tsv")
    _dump(full_path, regtest)

    # collect q-value stats
    m_score_stat, = [l for l in stdout.split("\n") if "mean q_value" in l]

    # split away log time etc:
    __, __, interesting_m_score_output = m_score_stat.partition("mean q_value")

    # collect s value stats
    s_value_stat, = [l for l in stdout.split("\n") if "mean s_value" in l]

    # split away log time etc:
    __, __, interesting_s_value_output = s_value_stat.partition("mean s_value")

    stdout = _run_cmdline("pyprophet score --in=test_data.txt --apply_weights=test_data_ms2_weights.csv "
                          "--test")

    _record(stdout, regtest)

    # collect m score stats
    m_score_stat, = [l for l in stdout.split("\n") if "mean q_value" in l]

    # split away log time etc:
    __, __, interesting_m_score_output2 = m_score_stat.partition("mean q_value")

    # collect s value stats
    s_value_stat, = [l for l in stdout.split("\n") if "mean s_value" in l]

    # split away log time etc:
    __, __, interesting_s_value_output2 = s_value_stat.partition("mean s_value")

    # check if output is consistent over runs:
    assert interesting_m_score_output == interesting_m_score_output2
    assert interesting_s_value_output == interesting_s_value_output2


def test_not_unique_tg_id_blocks(tmpdir):

    os.chdir(tmpdir.strpath)
    data_path = os.path.join(DATA_FOLDER, "test_invalid_data.txt")
    shutil.copy(data_path, tmpdir.strpath)
    cmdline = "pyprophet score --in=test_invalid_data.txt --test"

    with pytest.raises(subprocess.CalledProcessError) as exc_info:
        subprocess.check_output(cmdline, shell=True, stderr=subprocess.STDOUT)

    e = exc_info.value
    assert "Error: group_id values do not form unique blocks in input file(s)." in e.output
