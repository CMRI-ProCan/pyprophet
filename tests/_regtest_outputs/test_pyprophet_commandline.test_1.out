         0       1      2                               3                  4                            5                    6                              7                          8                 9   \
0  group_id  run_id  decoy  main_var_xx_swath_prelim_score  var_bseries_score  var_elution_model_fit_score  var_intensity_score  var_isotope_correlation_score  var_isotope_overlap_score  var_library_corr
1  459_run0       0      0                   5.27893963303                  5               0.993098149697       0.930480707464                 0.998724247981                        0.0     0.91240602174
2  459_run0       0      0                  -0.13974375122                  0               0.953924566507      0.0259317385401                 0.632511052964                        0.0   -0.285526965432
3  459_run0       0      0                  -0.77504285962                  0               0.666666249434     0.00044235703213                 0.657922811238                        0.0    0.376474234306
4  459_run0       0      0                  -1.05454511848                  0               0.847340305646      0.0009039469787                  0.79843565088             0.255319140851     0.51872168654
5  459_run0       0      0                    -1.246166932                  0                         -1.0     0.00042312411769                 0.563846670955            0.0454545468092    0.445906014642
6  459_run0       0      0                  -1.27951627249                  1               0.931647390127     0.00561216443373               -0.0419896690886            0.0997258424759  -0.0976738454584
7  459_run0       0      0                  -1.81518747764                  0               0.933585554361     0.00048082286101                 0.104774965663             0.159999996424   -0.630422997778
8  460_run0       0      0                   4.90480697454                  0               0.991616976261       0.929647682861                 0.998396899384                        0.0    0.757169232785
9  460_run0       0      0                   1.00012878546                  0               0.836714565754     0.00222462693417                 0.762015540501             0.380021706223    0.956728181447

        ...                          15                            16               17                        18                 19              20                21                 22                 23  \
0       ...         var_xcorr_coelution  var_xcorr_coelution_weighted  var_xcorr_shape  var_xcorr_shape_weighted  var_yseries_score         d_score           p_value            q_value                pep
1       ...                         0.0                           0.0   0.960350100659            0.970979341002                  6    5.5453286171  0.00258397932817  0.000230027833368  0.000719713388069
2       ...               1.39668108475                0.859180469403   0.789175310515            0.716396888252                  0  -1.22927618027    0.881136950904    0.0684821464433                1.0
3       ...               2.38678570295                 1.97001683336   0.119047619048           0.0149915833201                  0  -8.01535320282    0.984496124031    0.0763175289947                1.0
4       ...               4.39320066984                 2.78930134498   0.329194873727            0.330811834697                  0  -5.11298179626    0.984496124031    0.0763175289947                1.0
5       ...                         1.0                           1.0              0.0                       0.0                  0  -1.10749149323    0.881136950904    0.0684821464433                1.0
6       ...               2.20848900055                 1.49086544255   0.711876219914            0.685971015368                  1  -3.39510893822    0.984496124031    0.0763175289947                1.0
7       ...               1.39696341001                0.799699939913   0.761807409832            0.743328859302                  0  -4.13641691208    0.984496124031    0.0763175289947                1.0
8       ...               2.17321312465                 1.00181054984   0.895313998806            0.867887743561                  5   4.53198957443  0.00258397932817  0.000230027833368  0.000719713388069
9       ...               3.54384793756                 1.31714642152   0.666435350514            0.636692707326                  0  -1.17818796635    0.881136950904    0.0684821464433                1.0

                24
0  peak_group_rank
1                1
2                3
3                7
4                6
5                2
6                4
7                5
8                1
9                2

[10 rows x 25 columns]
...
                  0  1  2                3  4               5                 6                7                8                 9  ...             15              16              17              18 19  \
9156  DECOY_592_run0  0  1  -0.909754229584  0  0.823102881511   0.0126013214359  -0.147306602511   0.229729726911    0.778375379467 ...  3.94538540661   2.00615080523  0.604014629053  0.541176149999  2
9157  DECOY_592_run0  0  1  -0.956865429723  0            -1.0  0.00238403378516   0.266941597725              0.0  -0.0969535542793 ...            1.0             1.0             0.0             0.0  0
9158  DECOY_592_run0  0  1   -1.52418394133  0            -1.0  0.00017028812751  -0.583329727443              0.0  -0.0585549720413 ...            1.0             1.0             0.0             0.0  0
9159  DECOY_592_run0  0  1   -1.68998208875  0  0.895950257778  0.00854846400109   0.546651481853   0.601593613625   -0.189017878466 ...  2.06934963933   1.14626069677  0.785186892817  0.764795972809  0
9160  DECOY_592_run0  0  1     -1.692134624  0  0.958494941394  0.00630066071793  -0.328751834466   0.189189195633   -0.482215863506 ...  1.28809100151  0.791833291782  0.803144337956  0.768010655265  0
9161  DECOY_592_run0  0  1   -1.84945466301  0  0.633835911751  0.00561950820789   0.371593955459              0.0   -0.371771330281 ...  2.61135263418   2.02039942856   0.55597268955  0.330880632506  0
9162  DECOY_592_run0  0  1   -2.10957129817  0  0.906453510125  0.00527893195286    0.22376248428  0.0645161271095   -0.255514614107 ...  1.07777568371  0.598394985285  0.893420898864  0.890335674837  0
9163  DECOY_592_run0  0  1   -2.31560973872  0  0.905198852221   0.0028948981677   0.137664821691              0.0    0.876978922737 ...  3.43237660856   2.17985387492  0.539929223558   0.44766613662  0
9164  DECOY_592_run0  0  1   -2.53264045217  0  0.980351785819  0.00664123697296  0.0745622896143   0.205128207803    0.216855907728 ...  1.87132796282  0.908604380766  0.736385791197  0.708796452491  0
9165  DECOY_592_run0  0  1   -3.44062881815  0            -1.0  0.00102172876507  0.0339463918184   0.333333343267   0.0837371951856 ...            1.0             1.0             0.0             0.0  0

                   20              21               22   23  24
9156    -4.1786365509  0.984496124031  0.0763175289947  1.0  13
9157  -0.988372802734  0.870801033592  0.0680313307494  1.0   3
9158    -4.3621339798  0.984496124031  0.0763175289947  1.0  15
9159   -3.50024700165  0.984496124031  0.0763175289947  1.0  11
9160   -3.08865404129  0.984496124031  0.0763175289947  1.0   8
9161   -3.18250703812  0.984496124031  0.0763175289947  1.0   9
9162   -1.84614098072  0.984496124031  0.0763175289947  1.0   4
9163   -5.23752641678  0.984496124031  0.0763175289947  1.0  16
9164   -4.33319282532  0.984496124031  0.0763175289947  1.0  14
9165    -3.4147105217  0.984496124031  0.0763175289947  1.0  10

[10 rows x 25 columns]
                                                    0
0                                  score,weight,level
1   main_var_xx_swath_prelim_score,0.192501032323,ms2
2              var_bseries_score,-0.0728896042615,ms2
3      var_elution_model_fit_score,-4.37865033011,ms2
4              var_intensity_score,0.942967187402,ms2
5     var_isotope_correlation_score,1.74876561622,ms2
6        var_isotope_overlap_score,-2.01783122696,ms2
7                var_library_corr,-0.214427161054,ms2
8                 var_library_rmsd,-8.43977281165,ms2
9               var_log_sn_score,-0.0451398349059,ms2
10              var_massdev_score,0.0241725434287,ms2
11    var_massdev_score_weighted,-0.0123559669203,ms2
12               var_norm_rt_score,-18.3386295658,ms2
13             var_xcorr_coelution,0.197223243295,ms2
14   var_xcorr_coelution_weighted,-0.184838816428,ms2
15                  var_xcorr_shape,5.30844138363,ms2
16          var_xcorr_shape_weighted,4.1891085706,ms2
17               var_yseries_score,0.191171770248,ms2
                                                   0
0  qvalue,pvalue,svalue,pep,fdr,fnr,fpr,tp,tn,fp,...
1  0.0,0.00258397932817,0.943760449916,0.00071971...
2  0.00999999977648,0.113695090439,1.0,0.44172638...
3  0.019999999553,0.297157622739,1.0,1.0,0.024224...
4  0.0500000007451,0.617571059432,1.0,1.0,0.04888...
5  0.10000000149,0.984496124031,1.0,1.0,0.0763175...
6                           0.20000000298,,,,,,,,,,,
7                          0.300000011921,,,,,,,,,,,
8                           0.40000000596,,,,,,,,,,,
9                                     0.5,,,,,,,,,,,
                                                   0
0  cutoff,fdr,fn,fnr,fp,fpr,pvalue,qvalue,svalue,...
1  -2.3650124073,0.0763175289947,-0.46511627907,0...
2  -2.16003394127,0.0763175289947,-0.46511627907,...
3  -1.95505547523,0.0763175289947,-0.46511627907,...
4  -1.7500770092,0.0763175289947,-0.46511627907,0...
5  -1.54509854317,0.0763175289947,-0.46511627907,...
6  -1.34012007713,0.0684821464433,-2.56589147287,...
7  -1.13514149189,0.0684821464433,-2.56589147287,...
8  -0.930163025856,0.0680313307494,-0.87596899224...
9  -0.725184559822,0.0595057380533,-3.20930232558...
...
                                                    0
42  6.03910541534,0.000230027833368,20.0775193798,...
43  6.24408388138,0.000230027833368,20.0775193798,...
44  6.44906234741,0.000230027833368,20.0775193798,...
45  6.65404081345,0.000230027833368,20.0775193798,...
46  6.85901927948,0.000230027833368,20.0775193798,...
47  7.06399774551,0.000230027833368,20.0775193798,...
48  7.26897621155,0.000230027833368,20.0775193798,...
49  7.47395515442,0.000230027833368,20.0775193798,...
50  7.67893362045,0.000230027833368,20.0775193798,...
51  7.88391208649,0.000230027833368,20.0775193798,...
pyprophet score --in=test_data.txt --random_seed=42 --parametric
 config settings:
     final_statistics.emp_p: True
     final_statistics.lambda: 0.4
     final_statistics.lfdr_adj: 1.5
     final_statistics.lfdr_eps: 1e-08
     final_statistics.lfdr_monotone: True
     final_statistics.lfdr_transf: probit
     final_statistics.lfdr_trunc: True
     final_statistics.pfdr: False
     final_statistics.pi0_method: smoother
     final_statistics.pi0_smooth_df: 3
     final_statistics.pi0_smooth_log_pi0: False
     group_id: group_id
     is_test: False
     num_processes: 1
     random_seed: 42
     semi_supervised_learner.initial_fdr: 0.15
     semi_supervised_learner.initial_lambda: 0.4
     semi_supervised_learner.iteration_fdr: 0.02
     semi_supervised_learner.iteration_lambda: 0.4
     semi_supervised_learner.num_iter: 10
     tric_chromprob: False
     xeval.fraction: 0.5
     xeval.num_iter: 10
 learn and apply classifier from input data
 Data set contains 387 decoy and 387 target groups.
 summary input file:
    9165 lines
    774 transition groups
    17 scores including main score
 learn and apply scorer
 start 10 cross evals using 1 processes
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 finished cross evals

 Data set contains 387 decoy and 387 target groups.
 mean q_value = 6.731646e-02, std_dev q_value = 1.984248e-02
 mean s_value = 9.976264e-01, std_dev s_value = 1.112700e-02
 calculated scoring and statistics
 processing input data finished
==================================================================================================
   qvalue    pvalue   svalue       pep       fdr      fnr       fpr          tp         tn  \
0    0.00  0.002584  0.94376  0.000720  0.000230  0.40155  0.002584  336.922481  29.922481
1    0.01  0.113695  1.00000  0.441726  0.009396  0.00000  0.113695  359.589147  26.589147
2    0.02  0.297158  1.00000  1.000000  0.024225  0.00000  0.297158  359.085271  21.085271
3    0.05  0.617571  1.00000  1.000000  0.048884  0.00000  0.617571  360.472868  11.472868
4    0.10  0.984496  1.00000  1.000000  0.076318  0.00000  0.984496  357.465116   0.465116
5    0.20       NaN      NaN       NaN       NaN      NaN       NaN         NaN        NaN
6    0.30       NaN      NaN       NaN       NaN      NaN       NaN         NaN        NaN
7    0.40       NaN      NaN       NaN       NaN      NaN       NaN         NaN        NaN
8    0.50       NaN      NaN       NaN       NaN      NaN       NaN         NaN        NaN

          fp         fn    cutoff
0   0.077519  20.077519  3.030515
1   3.410853  -2.589147  1.193835
2   8.914729  -2.085271  0.369034
3  18.527132  -3.472868 -0.369136
4  29.534884  -0.465116 -1.899152
5        NaN        NaN       NaN
6        NaN        NaN       NaN
7        NaN        NaN       NaN
8        NaN        NaN       NaN
==================================================================================================
WRITTEN: test_data_ms2_summary_stat.csv
WRITTEN: test_data_ms2_full_stat.csv
WRITTEN: test_data_ms2_scored.tsv
WRITTEN: test_data_ms2_report.pdf
WRITTEN: test_data_ms2_weights.csv
TIME: 0 seconds and 854 msecs wall time

